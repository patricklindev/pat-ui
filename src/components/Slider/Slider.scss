@import './styles/slider_mixins';
@import './styles/slider_vars';


@include slider("small", "primary");
@include slider("medium", "primary");
@include slider("large", "primary");
@include slider("small", "secondary");
@include slider("medium", "secondary");
@include slider("large", "secondary");
@include slider("small", "disabled");
@include slider("medium", "disabled");
@include slider("large", "disabled");