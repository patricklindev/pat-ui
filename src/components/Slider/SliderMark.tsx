import React from "react";
// import { SliderSize, ButtonType } from "./Slider";
import { classNames } from "../../utils/classNames";

export const SliderMark = (props: any) => {
  const { sliderType, sliderSize, mark, position, active, disabled } = props;

  let styleMarkClasses = classNames('slider_mark', {
    [`slider_mark-${sliderType}`]: true,
    [`slider_mark-${sliderSize}`]: !!sliderSize,
    disabled: disabled,
  })

  let styleMarkMaskClasses = classNames('slider_mark-mask', {
    [`slider_mark-${sliderSize}`]: !!sliderSize,
    disabled: disabled,
  })

  return (
    <>
      <span className={styleMarkClasses} style={{ left: `${position}%` }} data-active={active} ></span>
      {!active && <span className={styleMarkMaskClasses} style={{ left: `${position}%` }}></span>}
      {!!mark.label && <span className={'slider_mark-label'} style={{ left: `${position}%` }} data-active={active} >{mark.label}</span>}
    </>
  )
}

export default SliderMark;