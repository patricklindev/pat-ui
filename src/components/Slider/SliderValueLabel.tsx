import React, { FC } from 'react';

export const SliderValueLabel: FC<{
  value: number;
  sliderType: string;
  sliderSize: string;
}> = ({ value, sliderType, sliderSize }) => {
  return (
    <span className={`slider_value-label`}>
      <span
        className={`slider_value-label-background slider_value-label-background-${sliderType} slider_value-label-background-${sliderSize}`}
      ></span>
      <span
        className={`slider_value-label-number slider_value-label-number-${sliderSize}`}
      >
        {value}
      </span>
    </span>
  );
};

export default SliderValueLabel;
