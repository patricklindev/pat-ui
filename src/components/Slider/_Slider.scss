.slider_container {
  width: 100%;
  position: relative;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  padding: 12px 0;

  &-labeled {
    margin-bottom: 1rem;
  }

  &-vertical {
    transform: rotate(-90deg);
    padding: 50% 0;
    writing-mode: vertical-lr;
  }
}

.slider {
  -webkit-appearance: none;
  appearance: none;
  background: transparent;
  cursor: pointer;
  width: 100%;
  padding: 12px 0;

  &::-webkit-slider-thumb {
    // -webkit-appearance: none;
    // appearance: none;
    visibility: hidden;
    border-radius: 50%;
    background: transparent;
  }

  &::-moz-range-thumb {
    border-radius: 50%;
    border: none;
    background: transparent;
  }

  &.disabled,
  &[disabled] {
    cursor: not-allowed;
    @include slider-type($slider-disabled-color);
  }

  &-sm {
    @include slider-size($slider-track-height-sm);
  }

  &-md {
    @include slider-size($slider-track-height-md);
  }

  &-primary {
    @include slider-type($slider-primary-main);
  }

  &-secondary {
    @include slider-type($slider-secondary-main);
  }
}

.slider_thumb {
  position: absolute;
  height: 20px;
  width: 20px;
  border-radius: 50%;
  cursor: pointer;
  transform: translateX(-50%);

  .disabled,
  &[data-disabled='true'] {
    pointer-events: none;
    background: $slider-disabled-color;
    width: $slider-thumb-width-disabled;
    height: $slider-thumb-height-disabled;
  }

  &-sm {
    width: $slider-thumb-width-sm;
    height: $slider-thumb-height-sm;
  }

  &-md {
    width: $slider-thumb-width-md;
    height: $slider-thumb-height-md;
  }

  &-primary {
    background: $slider-primary-main;
  }

  &-secondary {
    background: $slider-secondary-main;
  }
}

.slider_value-label-number {
  position: absolute;
  color: white;
  transform: translateX(-50%);
  left: 50%;
  text-align: center;
  border-radius: 50%;
  font-size: 0.75rem;

  &-sm {
    top: calc(-150% - 1.125rem);
  }

  &-md {
    top: calc(-100% - 1.125rem);
  }
}

.slider_value-label-background {
  position: absolute;
  width: 32px;
  height: 32px;
  left: calc(50% - 32px / 2);

  border-radius: 16px 16px 0px 16px;
  transform: rotate(45deg);

  &-sm {
    top: calc(-150% - 1.5rem);
  }

  &-md {
    top: calc(-100% - 1.5rem);
  }

  &-primary {
    background: $slider-primary-main;
  }

  &-secondary {
    background: $slider-secondary-main;
  }
}

.slider_value {
  pointer-events: none;
  position: absolute;
  top: 50%;
  transform: translateY(-50%);

  &.disabled,
  &[disabled] {
    background: $slider-disabled-color;
  }

  &-sm {
    height: $slider-track-height-sm;
    border-radius: $slider-track-height-sm / 2;
  }

  &-md {
    height: $slider-track-height-md;
    border-radius: $slider-track-height-md / 2;
  }

  &-primary {
    background: $slider-primary-main;
  }

  &-secondary {
    background: $slider-secondary-main;
  }
}

.slider_mark {
  position: absolute;
  top: 50%;
  transform: translate(-50%, -50%);

  &.disabled,
  &[disabled] {
    background-color: $slider-disabled-color;
  }

  &-sm {
    height: $slider-mark-height-sm;
    width: $slider-mark-width-sm;
  }

  &-md {
    height: $slider-mark-height-md;
    width: $slider-mark-width-md;
  }

  &-primary {
    background-color: $slider-primary-main;

    &[data-active='true'] {
      color: $slider-primary-dark;
    }
  }

  &-secondary {
    background-color: $slider-secondary-main;
  }

  &-mask {
    position: absolute;
    top: 50%;
    transform: translate(-50%, -50%);
    background-color: #ffffff80;

    &.disabled,
    &[disabled] {
      background-color: $slider-disabled-color;
    }
  }

  &-label {
    position: absolute;
    top: 50%;
    transform: translate(-50%, 25%);

    &-vertical {
      transform: translate(-50%, 100%);
    }
  }
}
