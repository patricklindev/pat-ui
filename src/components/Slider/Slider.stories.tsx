import React from 'react';
import { action } from '@storybook/addon-actions';
import Slider from './Slider';

export default {
  title: 'Slider',
  component: Slider,
};

export const DefaultSlider = () => (
  <div>
    <h3>default</h3>
    <Slider onChange={(event) => action('' + event.target.value)(event)} />
  </div>
);

export const DiffSizeSlider = () => (
  <div>
    <h3>sm</h3>
    <Slider sliderSize="sm" />
    <br />
    <h3>md</h3>
    <Slider sliderSize="md" />
    <br />
    <h3>lg</h3>
    <Slider sliderSize="lg" />
  </div>
);

export const DiffColorSlider = () => (
  <div>
    <h3>primary</h3>
    <Slider sliderColor="primary" />
    <br />
    <h3>secondary</h3>
    <Slider sliderColor="secondary" />
    <br />
    <h3>default</h3>
    <Slider sliderColor="default" />
  </div>
);

export const DiffLabelOptions = () => (
  <div>
    <h3>off</h3>
    <Slider displayValueLabel="off" />
    <br />
    <h3>on</h3>
    <Slider displayValueLabel="on" />
    <br />
    <h3>auto</h3>
    <Slider displayValueLabel="auto" />
    <br />
  </div>
);

export const DiffInputModes = () => (
  <div>
    <h3>{'Step is not defined (continuous input mode)'}</h3>
    <Slider displayValueLabel="on" />
    <br />
    <h3>{'Step is defined (discrete input mode)'}</h3>
    <Slider displayValueLabel="on" step={10} />
    <br />
  </div>
);

export const DiffMarkOptions = () => (
  <div>
    <h3>marks with default</h3>
    <Slider marks={true} displayValueLabel={'auto'} />
    <br />
    <h3>marks with steps</h3>
    <br />
    <Slider marks={true} step={25} displayValueLabel={'auto'} />
    <br />
    <h3>custom marks</h3>
    <Slider marks={[20, 60, 75]} displayValueLabel={'auto'} />

    <br />
    <h3>custom marks with non-default length</h3>
    <Slider marks={[20, 60, 75]} max={80} displayValueLabel={'auto'} />
  </div>
);

export const RangeSlider = () => (
  <div>
    <h3>Range Slider with 2 thumbs</h3>
    <Slider rangePositions={[25, 75]} displayValueLabel={'on'} />
    <br />
    <h3>Range Slider with 3 thumbs</h3>
    <Slider rangePositions={[30, 60, 90]} displayValueLabel={'on'} />
    <br />
  </div>
);

export const DisabledSlider = () => <Slider disabled={true} />;
