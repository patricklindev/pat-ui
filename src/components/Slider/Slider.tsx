import React, { useState, ChangeEvent, FC } from 'react';
import { classNames } from '../../utils/classNames';
import SliderMark from './SliderMark';

export type SliderSize = 'sm' | 'md';
export type ButtonType = 'primary' | 'secondary';

export type mark = {
  label?: string,
  value: number,
}

export interface SliderProps {
  className?: string;
  sliderSize?: string;
  sliderType?: string;
  defaultValue?: number;
  disabled?: boolean;
  min?: number;
  max?: number;
  marks?: mark[]
}

export const Slider: FC<SliderProps> = ({
  className,
  sliderSize = 'md',
  sliderType = 'primary',
  defaultValue = 50,
  min = 0,
  max = 100,
  disabled,
  marks,
  ...rest
}) => {
  const [value, setValue] = useState(defaultValue);
  const markLabeled = (marks: any): boolean => {
    for (let mark of marks) {
      if (!!mark.label) {
        return true;
      }
    }
    return false;
  }

  let containerClasses = classNames('slider_container', {
    ['slider_container-labeled']: !!marks && markLabeled(marks),
  })

  let styleClasses = classNames('slider', {
    [`slider-${sliderType}`]: true,
    [`slider-${sliderSize}`]: !!sliderSize,
    disabled: !!disabled,
  })

  let styleValueClasses = classNames('slider_value', {
    [`slider_value-${sliderType}`]: true,
    [`slider_value-${sliderSize}`]: !!sliderSize,
    disabled: !!disabled,
  })

  if (className) {
    styleClasses += ' ' + className;
  }

  const changeHandler = (e: ChangeEvent<HTMLInputElement>) => {
    setValue(Math.round(Number(e.target.value)));
  };

  const calculatePos = (value: number, min: number, max: number) => {
    return (value - min) * 100 / (max - min);
  }

  return (
    <span className={containerClasses}>
      <span className={styleValueClasses} style={{ width: `${calculatePos(value, min, max)}%`, left: '0%' }}></span>
      {marks?.map((mark: any, index: number) => {
        return (
          <SliderMark key={mark.value + index} sliderType={sliderType} sliderSize={sliderSize} mark={mark} position={calculatePos(mark.value, min, max)} active={mark.value <= value} disabled={disabled} />
        )
      })}
      <input type="range" min={min} max={max} value={value} className={styleClasses} onChange={changeHandler} disabled={disabled} />
    </span>
  )
}

export default Slider;