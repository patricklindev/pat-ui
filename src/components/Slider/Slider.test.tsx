import React from 'react';

import { render, fireEvent } from '@testing-library/react';


import Slider, { SliderComponentProps } from './Slider';

describe('Slider', () => {
    it('should match snapshot', () => {
        const { asFragment } = render(<Slider />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should render default slider', () => {
        const sliderProps: SliderComponentProps = {
            onChange: jest.fn(),
        };
        const wrapper = render(<Slider {...sliderProps} />);
        const sliderContainer = wrapper.getByTestId(
            'slider__container'
        ) as HTMLElement;
        const sliderInput = wrapper.getByTestId('slider__input') as HTMLElement;
        expect(sliderInput).toBeInTheDocument();
        expect(sliderInput.tagName).toBe('INPUT');
        expect(sliderContainer).toHaveClass('slider slider-default slider-md');
        expect(sliderInput).toHaveValue('50');
        expect(sliderProps.onChange).toHaveBeenCalledTimes(0);
        fireEvent.change(sliderInput, { target: { value: 20 } });
        expect(sliderInput).toHaveValue('20');
        expect(sliderProps.onChange).toHaveBeenCalledTimes(1);
    });


    it('should render correct slider based on different props', () => {
        const sliderProps: SliderComponentProps = {
            sliderColor: 'primary',
            sliderSize: 'sm',
            defaultValue: 10,
            onChange: jest.fn(),
        };
        const wrapper = render(<Slider {...sliderProps} />);
        const sliderContainer = wrapper.getByTestId(
            'slider__container'
        ) as HTMLElement;
        const sliderInput = wrapper.getByTestId('slider__input') as HTMLElement;
        expect(sliderInput).toBeInTheDocument();
        expect(sliderInput.tagName).toBe('INPUT');
        expect(sliderContainer).toHaveClass('slider slider-primary slider-sm');
        expect(sliderInput).toHaveValue('10');
        expect(sliderProps.onChange).toHaveBeenCalledTimes(0);
        fireEvent.change(sliderInput, { target: { value: 20 } });
        expect(sliderInput).toHaveValue('20');
        expect(sliderProps.onChange).toHaveBeenCalledTimes(1);
    });

    it('should render marks slider', () => {
        const sliderProps: SliderComponentProps = {
            marks: [20, 50, 80],
        };
        const wrapper = render(<Slider {...sliderProps} />);
        const sliderMarks = wrapper.queryAllByTestId('slider__marks') as HTMLElement[];
        expect(sliderMarks).toHaveLength(3);
        const sliderNumberMarks = wrapper.queryAllByTestId('slider__number__marks') as HTMLElement[];
        expect(sliderNumberMarks).toHaveLength(3);
        expect(sliderNumberMarks[0]).toHaveTextContent('20');
        expect(sliderNumberMarks[1]).toHaveTextContent('50');
        expect(sliderNumberMarks[2]).toHaveTextContent('80');
    });

    it('should render range slider', () => {
        const sliderProps: SliderComponentProps = {
            rangePositions: [20, 50],
            valueLabelDisplay: 'on',
            onChange: jest.fn(),
        };
        const wrapper = render(<Slider {...sliderProps} />);
        const sliderBubbles = wrapper.queryAllByTestId(
            'slider__bubble'
        ) as HTMLElement[];
        const sliderInput = wrapper.getByTestId('slider__input') as HTMLElement;
        expect(sliderBubbles).toHaveLength(2);
        expect(sliderBubbles[0]).toHaveTextContent('20');
        expect(sliderBubbles[1]).toHaveTextContent('50');
        expect(sliderInput).toBeInTheDocument();
        expect(sliderInput.tagName).toBe('INPUT');
        expect(sliderProps.onChange).toHaveBeenCalledTimes(0);
        fireEvent.change(sliderInput, { target: { value: 10 } });
        expect(sliderInput).toHaveValue('10');
        expect(sliderProps.onChange).toHaveBeenCalledTimes(1);

        expect(sliderBubbles[0]).toHaveTextContent('10');
        expect(sliderBubbles[1]).toHaveTextContent('50');

    });

    it('should render vertical slider', () => {
        const sliderDisabledProps: SliderComponentProps = {
            orientation: 'vertical',
        };
        const wrapper = render(<Slider {...sliderDisabledProps} />);
        const sliderContainer = wrapper.getByTestId(
            'slider__container'
        ) as HTMLElement;
        expect(sliderContainer).toHaveClass('slider slider-vertical');
    });

    it('should render disabled slider', () => {
        const sliderDisabledProps: SliderComponentProps = {
            disabled: true,
        };
        const wrapper = render(<Slider {...sliderDisabledProps} />);
        const sliderContainer = wrapper.getByTestId(
            'slider__container'
        ) as HTMLElement;
        const sliderInput = wrapper.getByTestId('slider__input') as HTMLElement;
        expect(sliderInput).toBeInTheDocument();
        expect(sliderInput.tagName).toBe('INPUT');
        expect(sliderContainer).toHaveClass('slider slider-disabled');
        expect(sliderInput).toBeDisabled();
    });

});