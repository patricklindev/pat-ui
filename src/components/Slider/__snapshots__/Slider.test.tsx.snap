// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="slider slider-default slider-md slider-horizontal"
    data-testid="slider__container"
  >
    <div
      class="slider__rail"
      data-testid="slider__rail"
    />
    <div
      class="slider__track "
      style="width: 50%;"
    >
      <div
        class="slider__thumb"
      />
    </div>
    <input
      data-testid="slider__input"
      max="100"
      min="0"
      type="range"
      value="50"
    />
  </div>
</DocumentFragment>
`;
